Change Datasets model to models

cdeil opened the pull request:
This PR is an attempt to fix #2102 by changing `dataset.model` to `dataset.model` on all our five dataset classes. If we do this change, we should do it now, before v0.15.

There's still two failing tests (see [here](https://gist.github.com/cdeil/3cb466921b87b79acd91dffb930916be)) and I didn't update the tutorials or docs yet. There's merge conflicts already. Probably I'll have to re-start, but this PR shows that the change can be done within an hour.

As far as I understand it, our codebase already has migrated to this design that `dataset.model` should always be a `SkyModels` (plural) object. So IMO this change is the logical consequence, easier to understand for users. The logic is to accept for convenience different objects in init and in the setter (usually `SkyModel`, `SkyModels` and `list` of `SkyModel`), but then in the `@models.setter` to have checks and converters that guarantee that `_models` will always be a `SkyModels`, so everywhere where we access it (~ 1000 places) we can rely on that.

There's still the concern that we allow `None` and empty `SkyModels` list, which can cause issues (see [here](https://gist.github.com/cdeil/3cb466921b87b79acd91dffb930916be)). We could either allow both and getters have to check on access. Or we only allow one in the setter.

@adonath @registerrier @QRemy - Thoughts?

If we go this way: should I rebase here and we try to push this in? Or should I re-start at a better time (e.g. Sunday evening)?

QRemy commented on gammapy/cube/fit.py:
```diff
@@ -804,7 +811,7 @@ def to_dict(self, filename=""):
             "type": self.tag,
             "likelihood": self.likelihood_type,
-            "models": [_.name for _ in self.model],
+            "models": [_.name for _ in self.models],
```
This fail for the None case no ?

cdeil replied:
Yes.
How should we fix?
Adapt this and other callers and add an if/else?
Or create `self._models = SkyModels([])` in `@models.setter`?

QRemy replied:
we have this in the spectrumDatasetOnOff
"""
        if self.model is not None:
            models = [_.name for _ in self.model]
        else:
            models = []

        return {
            "name": self.name,
            "type": self.tag,
            "models": models,
            "likelihood": self.likelihood_type,
            "filename": filename,
        }
"""

cdeil replied:
So I put the `if` and we allow `None`?
I'm not sure, but I think that would also be my preference.

cdeil replied:
I've put the same if in both places. @QRemy - Thanks for the advice!

adonath commented:
Thanks @cdeil, I'm +1 to rebase. In #2632 I removed the `e_true` argument from the `SpectrumEvaluator`, so you have to adapt the changes here.

Concerning `None` vs. an empty `SkyModels`. I don't really care. Both works, I guess the empty models container leads to less code, but as long as we don't have something like `SkyModels.append()` it's barely useful for users and becomes and implementation detail...do whatever you prefer.

cdeil commented:
@adonath - I've rebased and fixed all tests. I did not update the tutorials or docs yet.

I have to go now. If you can finish this up and merge it in, please do.
Otherwise, please leave a review and I'll try to do it tonight or tomorrow.

cdeil commented:
Just FYI: I would suggest to rename `SkyModels` -> `Models` tomorrow, and use that as a generic container, i.e. for special datasets like dark matter, they can add a model on the dataset anywhere they like with whatever parameters they like, and everything will work as long as they stick the models in `dataset.models`, it can be background or IRF or any other models.
Whether we manage to refactor BG models or not before v0.15 or v1.0, this would be my suggestion where to go.

adonath commented:
@cdeil I have some time now to work on this, I'll fix the remaining test fails and merge this PR after...

adonath commented:
Thanks @cdeil!

cdeil commented:
@adonath - Thank you for finishing this up!
